File: wsl2/wsl2_automation.py
#!/usr/bin/env python3
"""
WSL2 Two-Phase Automation Script
Handles the challenge of making changes in WSL2, then Windows, then restarting WSL2
"""

import os
import sys
import subprocess
import json
import time
import argparse
import shutil
from pathlib import Path
from typing import Dict, Any

class WSL2Automator:
    def __init__(self, 
                 kernel_branch="linux-msft-wsl-6.6.y", 
                 kernel_dest="C:/bzImage",
                 kernel_repo="https://github.com/microsoft/WSL2-Linux-Kernel.git",
                 kernel_dir="WSL2-Linux-Kernel",
                 auto_clone=True):
        self.is_wsl = self._check_wsl()
        self.windows_user = self._get_windows_user()
        self.temp_dir = Path("/tmp/wsl2_automation") if self.is_wsl else Path("C:/temp/wsl2_automation")
        self.state_file = self.temp_dir / "automation_state.json"
        self.kernel_branch = kernel_branch
        self.kernel_dest = kernel_dest
        self.kernel_repo = kernel_repo
        self.kernel_dir = kernel_dir
        self.auto_clone = auto_clone
        
    def _check_wsl(self) -> bool:
        """Check if running in WSL environment"""
        try:
            with open('/proc/version', 'r') as f:
                return 'microsoft' in f.read().lower()
        except:
            return False
    
    def _get_windows_user(self) -> str:
        """Get Windows username"""
        if self.is_wsl:
            try:
                result = subprocess.run(['cmd.exe', '/c', 'echo %USERNAME%'], 
                                      capture_output=True, text=True)
                return result.stdout.strip()
            except:
                return ""
        else:
            return os.environ.get('USERNAME', '')
    
    def _log(self, message: str, level: str = "INFO"):
        """Logging function with colors"""
        colors = {
            "INFO": "\033[0;34m",
            "SUCCESS": "\033[0;32m", 
            "WARNING": "\033[1;33m",
            "ERROR": "\033[0;31m"
        }
        reset = "\033[0m"
        timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
        print(f"{colors.get(level, '')}{timestamp} [{level}]{reset} {message}")
    
    def _save_state(self, state: Dict[str, Any]):
        """Save automation state to file"""
        self.temp_dir.mkdir(exist_ok=True)
        with open(self.state_file, 'w') as f:
            json.dump(state, f, indent=2)
    
    def _load_state(self) -> Dict[str, Any]:
        """Load automation state from file"""
        if self.state_file.exists():
            with open(self.state_file, 'r') as f:
                return json.load(f)
        return {}
    
    def _run_command(self, cmd: list, check: bool = True, capture_output: bool = False) -> subprocess.CompletedProcess:
        """Run command with logging"""
        self._log(f"Running: {' '.join(cmd)}")
        if capture_output:
            return subprocess.run(cmd, check=check, capture_output=True, text=True)
        return subprocess.run(cmd, check=check, text=True)
    
    def phase1_wsl_tasks(self):
        """Phase 1: Tasks to run inside WSL2"""
        if not self.is_wsl:
            self._log("Phase 1 must run inside WSL2", "ERROR")
            return False
            
        self._log("Starting Phase 1: WSL2 tasks")
        
        # Save current directory
        original_dir = os.getcwd()
        
        try:
            # Example WSL2 tasks (customize as needed)
            self._log("Building kernel...")
            
            # Clone and build kernel (simplified version of your script)
            if not Path(self.kernel_dir).exists():
                if self.auto_clone:
                    self._log(f"Cloning kernel repository from {self.kernel_repo}...")
                    self._run_command([
                        "git", "clone", self.kernel_repo, self.kernel_dir,
                        "--depth=1", "-b", self.kernel_branch
                    ])
                else:
                    self._log(f"Kernel directory '{self.kernel_dir}' not found!", "ERROR")
                    self._log(f"Please clone the repository manually or set auto_clone=True", "ERROR")
                    return False
            else:
                self._log(f"Using existing kernel directory: {self.kernel_dir}")
            
            os.chdir(self.kernel_dir)
            
            # Install dependencies
            self._run_command([
                "sudo", "apt", "update"
            ])
            self._run_command([
                "sudo", "apt", "install", "-y",
                "build-essential", "flex", "bison", "libssl-dev", 
                "libelf-dev", "bc", "python3", "pahole", "cpio"
            ])
            
            # Build kernel
            try:
                nproc_result = self._run_command(["nproc"], capture_output=True)
                nproc = nproc_result.stdout.strip() if nproc_result.stdout else "4"
            except:
                nproc = "4"  # fallback
            self._run_command([
                "make", f"-j{nproc}", "KCONFIG_CONFIG=Microsoft/config-wsl"
            ])
            
            # Install modules
            self._run_command(["sudo", "make", "modules_install", "headers_install"])
            
            # Copy kernel to Windows
            kernel_path = Path("arch/x86/boot/bzImage")
            if kernel_path.exists():
                self._run_command(["cp", str(kernel_path), self.kernel_dest])
                self._log(f"Kernel copied to {self.kernel_dest}", "SUCCESS")
            else:
                self._log("Kernel image not found! Build may have failed.", "ERROR")
                return False
            
            os.chdir("..")
            
            # Save state for Phase 2
            state = {
                "phase1_completed": True,
                "kernel_built": True,
                "windows_user": self.windows_user,
                "timestamp": time.time()
            }
            self._save_state(state)
            
            # Create Phase 2 script on Windows
            self._create_phase2_script()
            
            self._log("Phase 1 completed successfully", "SUCCESS")
            self._log("Starting Phase 2 on Windows...")
            
            # Trigger Phase 2 on Windows
            self._trigger_phase2()
            
            return True
            
        except subprocess.CalledProcessError as e:
            self._log(f"Command failed: {e}", "ERROR")
            return False
        except Exception as e:
            self._log(f"Phase 1 failed: {e}", "ERROR")
            return False
        finally:
            # Always return to original directory
            os.chdir(original_dir)
    
    def _create_phase2_script(self):
        """Create the Phase 2 PowerShell script on Windows"""
        if not self.windows_user:
            self._log("Cannot determine Windows user", "ERROR")
            self._log("Please set WINDOWS_USER environment variable or run from WSL2", "ERROR")
            raise RuntimeError("Windows user not detected")
            
        ps_script = f"""
# Phase 2: Windows tasks
$ErrorActionPreference = "Stop"

Write-Host "Starting Phase 2: Windows tasks" -ForegroundColor Blue

try {{
    # Create .wslconfig
    $wslConfigPath = "C:\\Users\\{self.windows_user}\\.wslconfig"
    $wslConfig = @"
[wsl2]
kernel={self.kernel_dest}
"@
    
    Write-Host "Creating WSL config at $wslConfigPath" -ForegroundColor Green
    $wslConfig | Out-File -FilePath $wslConfigPath -Encoding UTF8
    
    # Wait a moment
    Start-Sleep -Seconds 2
    
    # Shutdown WSL
    Write-Host "Shutting down WSL..." -ForegroundColor Yellow
    wsl --shutdown
    
    # Wait for shutdown
    Start-Sleep -Seconds 5
    
    # Restart WSL (optional - will happen automatically on next use)
    Write-Host "WSL automation completed successfully!" -ForegroundColor Green
    Write-Host "WSL will use the new kernel on next startup." -ForegroundColor Green
    
}} catch {{
    Write-Host "Phase 2 failed: $_" -ForegroundColor Red
    exit 1
}}

# Cleanup
Remove-Item "C:\\temp\\wsl2_automation\\phase2.ps1" -ErrorAction SilentlyContinue
"""
        
        # Write PowerShell script to Windows
        ps_script_path = Path("/mnt/c/temp/wsl2_automation")
        ps_script_path.mkdir(parents=True, exist_ok=True)
        
        with open(ps_script_path / "phase2.ps1", 'w') as f:
            f.write(ps_script)
        
        self._log("Phase 2 script created on Windows", "SUCCESS")
    
    def _trigger_phase2(self):
        """Trigger Phase 2 execution on Windows"""
        try:
            # Execute PowerShell script on Windows
            self._run_command([
                "powershell.exe", "-ExecutionPolicy", "Bypass",
                "-File", "C:\\temp\\wsl2_automation\\phase2.ps1"
            ])
        except subprocess.CalledProcessError:
            # This is expected as WSL will be shut down
            self._log("Phase 2 triggered (WSL shutdown expected)", "SUCCESS")
    
    def phase2_windows_tasks(self):
        """Phase 2: Tasks to run on Windows (fallback if called directly)"""
        if self.is_wsl:
            self._log("Phase 2 should run on Windows, not WSL", "ERROR")
            return False
            
        self._log("Starting Phase 2: Windows tasks")
        
        try:
            state = self._load_state()
            if not state.get("phase1_completed"):
                self._log("Phase 1 not completed", "ERROR")
                return False
            
            # Create .wslconfig
            wsl_config_path = Path(f"C:/Users/{self.windows_user}/.wslconfig")
            wsl_config_content = f"[wsl2]\nkernel={self.kernel_dest}\n"
            
            with open(wsl_config_path, 'w') as f:
                f.write(wsl_config_content)
            
            self._log(f"WSL config created at {wsl_config_path}", "SUCCESS")
            
            # Shutdown WSL
            self._log("Shutting down WSL...")
            self._run_command(["wsl", "--shutdown"])
            
            self._log("Phase 2 completed successfully", "SUCCESS")
            return True
            
        except Exception as e:
            self._log(f"Phase 2 failed: {e}", "ERROR")
            return False
    
    def cleanup(self):
        """Clean up temporary files"""
        try:
            if self.temp_dir.exists():
                shutil.rmtree(self.temp_dir)
            self._log("Cleanup completed", "SUCCESS")
        except Exception as e:
            self._log(f"Cleanup failed: {e}", "WARNING")

def main():
    parser = argparse.ArgumentParser(description="WSL2 Two-Phase Automation")
    parser.add_argument("--phase", choices=["1", "2"], default="1",
                       help="Which phase to run (default: 1)")
    parser.add_argument("--cleanup", action="store_true",
                       help="Clean up temporary files")
    parser.add_argument("--kernel-branch", default="linux-msft-wsl-6.6.y",
                       help="Kernel branch to build (default: linux-msft-wsl-6.6.y)")
    parser.add_argument("--kernel-dest", default="C:/bzImage",
                       help="Destination for kernel image (default: C:/bzImage)")
    parser.add_argument("--kernel-repo", 
                       default="https://github.com/microsoft/WSL2-Linux-Kernel.git",
                       help="Kernel repository URL")
    parser.add_argument("--kernel-dir", default="WSL2-Linux-Kernel",
                       help="Local kernel directory name (default: WSL2-Linux-Kernel)")
    parser.add_argument("--no-clone", action="store_true",
                       help="Don't auto-clone the repository")
    
    args = parser.parse_args()
    
    automator = WSL2Automator(
        kernel_branch=args.kernel_branch,
        kernel_dest=args.kernel_dest,
        kernel_repo=args.kernel_repo,
        kernel_dir=args.kernel_dir,
        auto_clone=not args.no_clone
    )
    
    if args.cleanup:
        automator.cleanup()
        return
    
    if args.phase == "1":
        success = automator.phase1_wsl_tasks()
    else:
        success = automator.phase2_windows_tasks()
    
    sys.exit(0 if success else 1)

if __name__ == "__main__":
    main() 

File: wsl2/wsl2_automation.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

// Colors for output
const (
	ColorReset  = "\033[0m"
	ColorRed    = "\033[31m"
	ColorGreen  = "\033[32m"
	ColorYellow = "\033[33m"
	ColorBlue   = "\033[34m"
)

// AutomationState represents the state passed between phases
type AutomationState struct {
	Phase1Completed bool   `json:"phase1_completed"`
	KernelBuilt     bool   `json:"kernel_built"`
	WindowsUser     string `json:"windows_user"`
	Timestamp       int64  `json:"timestamp"`
}

// WSL2Automator handles the two-phase automation
type WSL2Automator struct {
	IsWSL        bool
	WindowsUser  string
	TempDir      string
	StateFile    string
	KernelBranch string
	KernelDest   string
	KernelRepo   string
	KernelDir    string
	AutoClone    bool
}

// NewWSL2Automator creates a new automator instance
func NewWSL2Automator(kernelBranch, kernelDest, kernelRepo, kernelDir string, autoClone bool) *WSL2Automator {
	automator := &WSL2Automator{
		KernelBranch: kernelBranch,
		KernelDest:   kernelDest,
		KernelRepo:   kernelRepo,
		KernelDir:    kernelDir,
		AutoClone:    autoClone,
	}
	automator.IsWSL = automator.checkWSL()
	automator.WindowsUser = automator.getWindowsUser()

	if automator.IsWSL {
		automator.TempDir = "/tmp/wsl2_automation"
	} else {
		automator.TempDir = "C:/temp/wsl2_automation"
	}

	automator.StateFile = filepath.Join(automator.TempDir, "automation_state.json")
	return automator
}

// checkWSL determines if running in WSL environment
func (w *WSL2Automator) checkWSL() bool {
	if runtime.GOOS != "linux" {
		return false
	}

	data, err := os.ReadFile("/proc/version")
	if err != nil {
		return false
	}

	return strings.Contains(strings.ToLower(string(data)), "microsoft")
}

// getWindowsUser retrieves the Windows username
func (w *WSL2Automator) getWindowsUser() string {
	if w.IsWSL {
		cmd := exec.Command("cmd.exe", "/c", "echo %USERNAME%")
		output, err := cmd.Output()
		if err != nil {
			return ""
		}
		return strings.TrimSpace(string(output))
	}
	return os.Getenv("USERNAME")
}

// log prints colored log messages
func (w *WSL2Automator) log(message, level string) {
	colors := map[string]string{
		"INFO":    ColorBlue,
		"SUCCESS": ColorGreen,
		"WARNING": ColorYellow,
		"ERROR":   ColorRed,
	}

	color := colors[level]
	if color == "" {
		color = ColorBlue
	}

	timestamp := time.Now().Format("2006-01-02 15:04:05")
	fmt.Printf("%s%s [%s]%s %s\n", color, timestamp, level, ColorReset, message)
}

// saveState saves automation state to file
func (w *WSL2Automator) saveState(state AutomationState) error {
	err := os.MkdirAll(w.TempDir, 0755)
	if err != nil {
		return err
	}

	data, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile(w.StateFile, data, 0644)
}

// loadState loads automation state from file
func (w *WSL2Automator) loadState() (AutomationState, error) {
	var state AutomationState

	data, err := os.ReadFile(w.StateFile)
	if err != nil {
		return state, err
	}

	err = json.Unmarshal(data, &state)
	return state, err
}

// runCommand executes a command with logging
func (w *WSL2Automator) runCommand(name string, args ...string) error {
	w.log(fmt.Sprintf("Running: %s %s", name, strings.Join(args, " ")), "INFO")
	cmd := exec.Command(name, args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

// runCommandOutput executes a command and returns output
func (w *WSL2Automator) runCommandOutput(name string, args ...string) (string, error) {
	w.log(fmt.Sprintf("Running: %s %s", name, strings.Join(args, " ")), "INFO")
	cmd := exec.Command(name, args...)
	output, err := cmd.Output()
	return strings.TrimSpace(string(output)), err
}

// phase1WSLTasks performs tasks inside WSL2
func (w *WSL2Automator) phase1WSLTasks() error {
	if !w.IsWSL {
		w.log("Phase 1 must run inside WSL2", "ERROR")
		return fmt.Errorf("not running in WSL2")
	}

	w.log("Starting Phase 1: WSL2 tasks", "INFO")

	// Save current directory
	originalDir, err := os.Getwd()
	if err != nil {
		return fmt.Errorf("failed to get current directory: %v", err)
	}
	// Ensure we return to original directory
	defer func() {
		if err := os.Chdir(originalDir); err != nil {
			w.log(fmt.Sprintf("Failed to return to original directory: %v", err), "WARNING")
		}
	}()

	// Check if kernel directory exists
	if _, err := os.Stat(w.KernelDir); os.IsNotExist(err) {
		if w.AutoClone {
			w.log(fmt.Sprintf("Cloning kernel repository from %s...", w.KernelRepo), "INFO")
			err := w.runCommand("git", "clone",
				w.KernelRepo, w.KernelDir,
				"--depth=1", "-b", w.KernelBranch)
			if err != nil {
				return fmt.Errorf("failed to clone kernel: %v", err)
			}
		} else {
			w.log(fmt.Sprintf("Kernel directory '%s' not found!", w.KernelDir), "ERROR")
			w.log("Please clone the repository manually or set --auto-clone", "ERROR")
			return fmt.Errorf("kernel directory not found and auto-clone disabled")
		}
	} else {
		w.log(fmt.Sprintf("Using existing kernel directory: %s", w.KernelDir), "INFO")
	}

	// Change to kernel directory
	err = os.Chdir(w.KernelDir)
	if err != nil {
		return fmt.Errorf("failed to change directory: %v", err)
	}

	// Install dependencies
	w.log("Installing build dependencies...", "INFO")
	err = w.runCommand("sudo", "apt", "update")
	if err != nil {
		return fmt.Errorf("failed to update packages: %v", err)
	}

	err = w.runCommand("sudo", "apt", "install", "-y",
		"build-essential", "flex", "bison", "libssl-dev",
		"libelf-dev", "bc", "python3", "pahole", "cpio")
	if err != nil {
		return fmt.Errorf("failed to install dependencies: %v", err)
	}

	// Build kernel
	w.log("Building kernel...", "INFO")
	nproc, err := w.runCommandOutput("nproc")
	if err != nil {
		nproc = "4" // fallback
	}

	err = w.runCommand("make", fmt.Sprintf("-j%s", nproc), "KCONFIG_CONFIG=Microsoft/config-wsl")
	if err != nil {
		return fmt.Errorf("failed to build kernel: %v", err)
	}

	// Install modules
	w.log("Installing kernel modules...", "INFO")
	err = w.runCommand("sudo", "make", "modules_install", "headers_install")
	if err != nil {
		return fmt.Errorf("failed to install modules: %v", err)
	}

	// Copy kernel to Windows
	kernelPath := "arch/x86/boot/bzImage"
	if _, err := os.Stat(kernelPath); err == nil {
		err = w.runCommand("cp", kernelPath, w.KernelDest)
		if err != nil {
			return fmt.Errorf("failed to copy kernel: %v", err)
		}
		w.log(fmt.Sprintf("Kernel copied to %s", w.KernelDest), "SUCCESS")
	} else {
		return fmt.Errorf("kernel image not found at %s - build may have failed", kernelPath)
	}

	// Save state for Phase 2
	state := AutomationState{
		Phase1Completed: true,
		KernelBuilt:     true,
		WindowsUser:     w.WindowsUser,
		Timestamp:       time.Now().Unix(),
	}

	err = w.saveState(state)
	if err != nil {
		return fmt.Errorf("failed to save state: %v", err)
	}

	// Create Phase 2 script
	err = w.createPhase2Script()
	if err != nil {
		return fmt.Errorf("failed to create phase 2 script: %v", err)
	}

	w.log("Phase 1 completed successfully", "SUCCESS")
	w.log("Starting Phase 2 on Windows...", "INFO")

	// Trigger Phase 2
	w.triggerPhase2()

	return nil
}

// createPhase2Script creates the PowerShell script for Phase 2
func (w *WSL2Automator) createPhase2Script() error {
	if w.WindowsUser == "" {
		w.log("Cannot determine Windows user", "WARNING")
		return nil
	}

	psScript := fmt.Sprintf(`# Phase 2: Windows tasks
$ErrorActionPreference = "Stop"

Write-Host "Starting Phase 2: Windows tasks" -ForegroundColor Blue

try {
    # Create .wslconfig
    $wslConfigPath = "C:\Users\%s\.wslconfig"
    $wslConfig = @"
[wsl2]
kernel=%s
"@
    
    Write-Host "Creating WSL config at $wslConfigPath" -ForegroundColor Green
    $wslConfig | Out-File -FilePath $wslConfigPath -Encoding UTF8
    
    # Wait a moment
    Start-Sleep -Seconds 2
    
    # Shutdown WSL
    Write-Host "Shutting down WSL..." -ForegroundColor Yellow
    wsl --shutdown
    
    # Wait for shutdown
    Start-Sleep -Seconds 5
    
    # Success message
    Write-Host "WSL automation completed successfully!" -ForegroundColor Green
    Write-Host "WSL will use the new kernel on next startup." -ForegroundColor Green
    
} catch {
    Write-Host "Phase 2 failed: $_" -ForegroundColor Red
    exit 1
}

# Cleanup
Remove-Item "C:\temp\wsl2_automation\phase2.ps1" -ErrorAction SilentlyContinue
`, w.WindowsUser, w.KernelDest)

	// Create directory on Windows
	psScriptDir := "/mnt/c/temp/wsl2_automation"
	err := os.MkdirAll(psScriptDir, 0755)
	if err != nil {
		return err
	}

	// Write PowerShell script
	psScriptPath := filepath.Join(psScriptDir, "phase2.ps1")
	err = os.WriteFile(psScriptPath, []byte(psScript), 0644)
	if err != nil {
		return err
	}

	w.log("Phase 2 script created on Windows", "SUCCESS")
	return nil
}

// triggerPhase2 executes the Phase 2 script on Windows
func (w *WSL2Automator) triggerPhase2() {
	// Execute PowerShell script on Windows
	cmd := exec.Command("powershell.exe", "-ExecutionPolicy", "Bypass",
		"-File", "C:\\temp\\wsl2_automation\\phase2.ps1")

	// This is expected to fail as WSL will be shut down
	cmd.Run()
	w.log("Phase 2 triggered (WSL shutdown expected)", "SUCCESS")
}

// phase2WindowsTasks performs tasks on Windows
func (w *WSL2Automator) phase2WindowsTasks() error {
	if w.IsWSL {
		w.log("Phase 2 should run on Windows, not WSL", "ERROR")
		return fmt.Errorf("running in WSL2")
	}

	w.log("Starting Phase 2: Windows tasks", "INFO")

	// Load state
	state, err := w.loadState()
	if err != nil || !state.Phase1Completed {
		w.log("Phase 1 not completed", "ERROR")
		return fmt.Errorf("phase 1 not completed")
	}

	// Create .wslconfig
	wslConfigPath := fmt.Sprintf("C:/Users/%s/.wslconfig", w.WindowsUser)
	wslConfigContent := fmt.Sprintf("[wsl2]\nkernel=%s\n", w.KernelDest)

	err = os.WriteFile(wslConfigPath, []byte(wslConfigContent), 0644)
	if err != nil {
		return fmt.Errorf("failed to create .wslconfig: %v", err)
	}

	w.log(fmt.Sprintf("WSL config created at %s", wslConfigPath), "SUCCESS")

	// Shutdown WSL
	w.log("Shutting down WSL...", "INFO")
	err = w.runCommand("wsl", "--shutdown")
	if err != nil {
		return fmt.Errorf("failed to shutdown WSL: %v", err)
	}

	w.log("Phase 2 completed successfully", "SUCCESS")
	return nil
}

// cleanup removes temporary files
func (w *WSL2Automator) cleanup() error {
	err := os.RemoveAll(w.TempDir)
	if err != nil {
		w.log(fmt.Sprintf("Cleanup failed: %v", err), "WARNING")
		return err
	}
	w.log("Cleanup completed", "SUCCESS")
	return nil
}

func main() {
	var phase = flag.String("phase", "1", "Which phase to run (1 or 2)")
	var cleanup = flag.Bool("cleanup", false, "Clean up temporary files")
	var kernelBranch = flag.String("kernel-branch", "linux-msft-wsl-6.6.y", "Kernel branch to build")
	var kernelDest = flag.String("kernel-dest", "C:/bzImage", "Destination for kernel image")
	var kernelRepo = flag.String("kernel-repo", "https://github.com/microsoft/WSL2-Linux-Kernel.git", "Kernel repository URL")
	var kernelDir = flag.String("kernel-dir", "WSL2-Linux-Kernel", "Local kernel directory name")
	var noClone = flag.Bool("no-clone", false, "Don't auto-clone the repository")
	flag.Parse()

	automator := NewWSL2Automator(*kernelBranch, *kernelDest, *kernelRepo, *kernelDir, !*noClone)

	if *cleanup {
		automator.cleanup()
		return
	}

	var err error
	switch *phase {
	case "1":
		err = automator.phase1WSLTasks()
	case "2":
		err = automator.phase2WindowsTasks()
	default:
		fmt.Printf("Invalid phase: %s\n", *phase)
		os.Exit(1)
	}

	if err != nil {
		automator.log(fmt.Sprintf("Automation failed: %v", err), "ERROR")
		os.Exit(1)
	}
}
